midend/lib/Utils/DIPUtils.cpp

```diff
@@ -410,6 +411,185 @@ Value customTanVal(OpBuilder &builder, Location loc, Value angleVal) {
+
+  Value strideVal = builder.create<arith::ConstantIndexOp>(loc, stride);
+
+  // generate vector [0., 1., ..., stride - 1]
+  std::vector<float> xInit(stride);
+  std::iota(xInit.begin(), xInit.end(), .0);
+  Value xVecInitial = builder.create<arith::ConstantOp>(
+      loc,
+      DenseFPElementsAttr::get(VectorType::get(stride, FloatType::getF32(ctx)),
+                               ArrayRef<float>(xInit)));
+
+  builder.create<scf::ParallelOp>(
+      loc,
+      /*lowerBounds=*/ValueRange{c0, c0},
+      /*upperBounds=*/ValueRange{outputRow, outputCol},
+      /*steps=*/ValueRange{c1, strideVal},
+      /*bodyBuilderFn=*/
+      [&](OpBuilder &builderFn, Location locFn, ValueRange ivsFn) {
+        Value yVec = builderFn.create<vector::SplatOp>(
+            locFn, vectorTyF32, indexToF32(builderFn, locFn, ivsFn[0]));
+        Value delta = builderFn.create<vector::SplatOp>(locFn, vectorTyF32, indexToF32(builderFn, locFn, ivsFn[1]));
+        Value xVec = builderFn.create<arith::AddFOp>(locFn, xVecInitial, delta);
+
+        // sourceXVec=(xVec*m0Vec*1024+(yVec*m1Vec+m2Vec)*1024+512)/1024
+        // sourceYVec=(xVec*m3Vec*1024+(yVec*m4Vec+m5Vec)*1024+512)/1024
+        auto getRes = [&](
+                          /*m0Vec or m3Vec*/ Value v1,
+                          /*m1Vec or m4Vec*/ Value v2,
+                          /*m2Vec or m5Vec*/ Value v3) {
+          Value x1 = builderFn.create<arith::MulFOp>(locFn, xVec, v1);
+          Value x2 = builderFn.create<arith::MulFOp>(locFn, x1, c1024F32Vec);
+          Value y1 = builderFn.create<vector::FMAOp>(locFn, yVec, v2, v3);
+          Value y2 = builderFn.create<arith::MulFOp>(locFn, y1, c1024F32Vec);
+          Value z1 = builderFn.create<arith::AddFOp>(locFn, x2, y2);
+          Value z2 = builderFn.create<arith::AddFOp>(locFn, z1, c512F32Vec);
+          return builderFn.create<arith::DivFOp>(locFn, z2, c1024F32Vec);
+        };
+        Value inputXVec = getRes(m0Vec, m1Vec, m2Vec);
+        Value inputYVec = getRes(m3Vec, m4Vec, m5Vec);
+
+        builderFn.create<AffineForOp>(
+            loc, ValueRange{c0}, builder.getDimIdentityMap(),
+            ValueRange{strideVal}, builder.getDimIdentityMap(), 1,
+            std::nullopt,
+            [&](OpBuilder &builderFor, Location locFor, ValueRange ivsFor,
+                ValueRange iterArg) {
+              Value inputPixelX = builderFor.create<vector::ExtractElementOp>(locFor, inputXVec, ivsFor[0]);
+              Value inputPixelY = builderFor.create<vector::ExtractElementOp>(locFor, inputYVec, ivsFor[0]);
+              Value inputPixelXI = F32ToIndex(builderFor, locFor,
+                                              builderFor.create<math::RoundOp>(locFor, inputPixelX));
+              Value inputPixelYI = F32ToIndex(builderFor, locFor,
+                                              builderFor.create<math::RoundOp>(locFor, inputPixelY));
+
+              auto inBound = [&](Value lb, Value ub, Value val) {
```

Comment:
Perhaps consider using [this generic](https://github.com/buddy-compiler/buddy-mlir/blob/cc57e60f073d769224a334b4df10c16b43eb2711/midend/lib/Utils/Utils.cpp#L95) function? Feel free to propose a change in its implementation if you want to.

Reply:
Function ValBound returns a Value, while this function returns boolean.
ValBound: return 0 if val < 0, val if 0 <= val <= lastElemf32, lastElemf32 if lastElemf32 < val
inBound: return true if lb <= val <= ub, false if val < lb or ub < val

Reply:
I think the intent at both places is similar i.e ensuring that the estimated pixel co-ordinates do not exceed output image dimensions. The function which you wrote does this by filtering pixels who have their estimated co-ordinates outside the output image, while `valBound` does this by clamping the dimension exceeding values to minimal/maximal allowed output image dimension value.

Reply:
Yes, but these two functions are not interchangeable

Reply:
Yes, but can you achieve the same functionality from both of them? 

Again, my purpose here is code reuse since pixel bounding is also used in resizing. If you think you need to change the implementation of `valBound` itself, feel free to propose a change and I'll review it.